File: 백준/Gold/2042. 구간 합 구하기/구간 합 구하기.cc
#include <bits/stdc++.h>
using namespace std;

int N, M, K;
long long leaf[1000001];
long long tree[1000001 * 4];

long long init(int start, int end, int node) {
	if(start == end) return tree[node] = leaf[start];
	
	int mid = (start + end) / 2;
	return tree[node] = init(start, mid, node * 2) + init(mid + 1, end, node * 2 + 1);
}

//left ~ right 합 구하는 구간 
long long prefixSum (int start, int end, int node, int left, int right) { 
	if(left > end || right < start) return 0;
	if(left <= start && end <= right) return tree[node];
	
	int mid = (start + end) / 2;
	return prefixSum(start, mid, node * 2, left, right) + prefixSum(mid + 1, end, node * 2 + 1, left, right); 
}

void update(int start, int end, int node, int index, long long diff) {
	if(index < start || index > end) return;
	
	tree[node] += diff;
	if(start == end) return;
	
	int mid = (start + end) / 2;
	update(start, mid, node * 2, index, diff);
	update(mid + 1, end, node * 2 + 1, index, diff);
} 

int main() {
	ios_base::sync_with_stdio(false); 
	cin.tie(NULL); 
	cout.tie(NULL);
	
	cin >> N >> M >> K;
	
	for(int i=1; i<=N; i++) {
		cin >> leaf[i];
	}

	init(1, N, 1);
	
	long long a, b, c;
	for(int i = 0; i < M + K; i++) {
		cin >> a >> b >> c;
		if(a == 1) {
			long long diff = c - leaf[b];
			leaf[b] = c;
			update(1, N, 1, b, diff);
		} else {
			cout << prefixSum(1, N, 1, b, c) << endl;
		}
	}
}

File: 백준/Gold/2042. 구간 합 구하기/구간 합 구하기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {

    static long init(int start, int end, int node, long[] tree, long[] A) {
        if (start == end) return tree[node] = A[start];
        int mid = start + (end - start) / 2;
        return tree[node] = init(start, mid, node * 2, tree, A) +
                init(mid + 1, end , node * 2 + 1, tree, A);
    }

    static long sum (int start, int end, int node, long left, long right, long[] tree) {
        if (left > end || right < start) return 0;
        if (left <= start && end <= right) return tree[node];
        int mid = start + (end - start) / 2;
        return sum(start, mid, node * 2, left, right, tree) +
                sum(mid + 1, end,  node * 2 + 1, left, right, tree);
    }

    static void update(int start, int end, int node, long changeNode, long diff, long[] tree) {
        if (start > changeNode || changeNode > end) return;

        tree[node] += diff;
        if (start == end) return;

        int mid = start + (end - start) / 2;
        update(start, mid, node * 2, changeNode, diff, tree);
        update(mid + 1, end, node * 2 + 1, changeNode, diff, tree);
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        StringBuilder result = new StringBuilder();
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());
        int k = Integer.parseInt(st.nextToken());

        long[] A = new long[n + 1];
        long[] tree = new long[(n + 1) * 4];
        for (int i = 1; i <= n; i++) {
            A[i] = Long.parseLong(br.readLine());
        }

        init(1, n, 1, tree, A);
        for (int i = 0; i < m + k; i++) {
            st = new StringTokenizer(br.readLine());
            long a = Long.parseLong(st.nextToken());
            long b = Long.parseLong(st.nextToken());
            long c = Long.parseLong(st.nextToken());

            // 업데이트
            if (a == 1) {
                long diff = c - A[(int) b];
                A[(int) b] = c;
                update(1, n, 1, b, diff, tree);
            }
            // 구간 합 구하기 [b, c]
            else if (a == 2) {
                result.append(sum(1, n, 1, b, c, tree)).append("\n");
            }
        }

        System.out.println(result.toString());
        br.close();
    }

}
